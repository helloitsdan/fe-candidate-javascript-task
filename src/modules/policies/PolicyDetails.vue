<template>
  <div>
    <div class="background--grey-green">
      <div v-if="!policy" class="container padded-block">
        <div v-if="loading" class="loading">Loading...</div>

        <div v-if="error" class="error">
          Sorry! We couldn't load this policy's details right now.
        </div>
      </div>

      <section v-if="policy" class="container padded-block">
        <nav>
          <p>
            <router-link :to="{ name: 'CustomerDetails' }">
              Back to policy list
            </router-link>
          </p>
        </nav>

        <h2>
          Policy for {{ policy.insured_entities | summariseInsuredEntities }}
        </h2>

        <policy-information v-bind="policy" />
      </section>
    </div>

    <div class="background--barely">
      <section v-if="policy" class="container padded-block">
        <policy-entities v-bind="policy" />
      </section>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import PolicyInformation from "@/components/PolicyInformation.vue";
import PolicyEntities from "@/components/PolicyEntities.vue";

import getPolicies from "@/services/policies";

export default Vue.extend({
  name: "PolicyDetails",
  props: ["policyId"],
  components: {
    PolicyInformation,
    PolicyEntities
  },
  data() {
    return {
      loading: false,
      error: null,
      policy: null
    };
  },
  created() {
    this.fetchPolicyDetails();
  },
  methods: {
    async fetchPolicyDetails() {
      const { policyId } = this.$props;

      if (this.loading) {
        return;
      }

      this.loading = true;

      try {
        this.policy = await getPolicies(policyId);
      } catch (e) {
        this.error = e;
      }

      this.loading = false;
    }
  }
});
</script>
