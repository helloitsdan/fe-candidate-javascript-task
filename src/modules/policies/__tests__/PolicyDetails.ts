import Vue from "vue";
import { shallowMount } from "@vue/test-utils";

import PolicyDetails from "../PolicyDetails.vue";
import policiesRequest from "@/services/policies";

jest.mock("@/services/policies");

describe("PolicyDetails", () => {
  const policiesRequestMock = policiesRequest as jest.Mock;

  const fakePolicy = {
    uuid: "1234",
    status: "ACTIVE",
    start_date: "2016-07-13T02:57:55.235Z",
    insured_entities: [
      { species: "CAT" },
      { species: "CAT" },
      { species: "DOG" }
    ],
  };

  it("renders", async () => {
    policiesRequestMock.mockReturnValue(Promise.resolve(fakePolicy));

    const wrapper = shallowMount(PolicyDetails, {
      stubs: ["router-link"]
    });

    await Vue.nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders a loading message when the request is still pending", () => {
    const wrapper = shallowMount(PolicyDetails, {
      stubs: ["router-link"]
    });
    expect(wrapper.find(".loading").exists()).toBe(true);
  });

  it("renders an error message when the request fails", async () => {
    policiesRequestMock.mockReturnValue(Promise.reject(":("));
    const wrapper = shallowMount(PolicyDetails, {
      stubs: ["router-link"]
    });

    await Vue.nextTick();

    expect(wrapper.find(".error").exists()).toBe(true);
  });
});
