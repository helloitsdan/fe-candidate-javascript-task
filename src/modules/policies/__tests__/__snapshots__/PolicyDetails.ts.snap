// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyDetails renders 1`] = `
<div>
  <div
    class="background--grey-green"
  >
    <!---->
     
    <section
      class="container padded-block"
    >
      <nav>
        <p>
          <router-link-stub
            to="[object Object]"
          >
            
            Back to policy list
          
          </router-link-stub>
        </p>
      </nav>
       
      <h2>
        
        Policy for 2 cats, 1 dog
      
      </h2>
       
      <policy-information-stub
        insured_entities="[object Object],[object Object],[object Object]"
        start_date="2016-07-13T02:57:55.235Z"
        status="ACTIVE"
        uuid="1234"
      />
    </section>
  </div>
   
  <div
    class="background--clay-dark"
  >
    <section
      class="container padded-block"
    >
      <policy-product-stub />
    </section>
  </div>
   
  <div
    class="background--barely"
  >
    <section
      class="container padded-block"
    >
      <policy-entities-stub
        insured_entities="[object Object],[object Object],[object Object]"
        start_date="2016-07-13T02:57:55.235Z"
        status="ACTIVE"
        uuid="1234"
      />
    </section>
  </div>
</div>
`;
